gnome-image-installer/pages/install/gis-scribe.c

```diff
@@ -0,0 +1,360 @@
+{
+  GTask *task = G_TASK (data);
+  GisScribe *self = GIS_SCRIBE (g_task_get_source_object (task));
+  g_autofree gchar *line = NULL;
+  g_auto(GStrv) arr = NULL;
+  gdouble curr, full;
+
+  if (g_io_channel_read_line (source, &line, NULL, NULL, NULL) != G_IO_STATUS_NORMAL
+      || line == NULL)
+    return G_SOURCE_CONTINUE;
+
+  if (!g_str_has_prefix (line, "[GNUPG:] PROGRESS"))
+    return G_SOURCE_CONTINUE;
+
+  /* https://git.gnupg.org/cgi-bin/gitweb.cgi?p=gnupg.git;a=blob;f=doc/DETAILS;h=0be55f4d;hb=refs/heads/master#l1043
+   * [GNUPG:] PROGRESS <what> <char> <cur> <total> [<units>]
+   * For example:
+   * [GNUPG:] PROGRESS /dev/mapper/endless- ? 676 4442 MiB
+   */
+  arr = g_strsplit (line, " ", -1);
+  curr = g_ascii_strtod (arr[4], NULL);
+  full = g_ascii_strtod (arr[5], NULL);
+
+  if (full == 0)
+    {
+      /* gpg can't determine the file size. Should not happen, since we
+       * gave it a hint.
+       */
+      self->progress = -1;
+    }
+  else
+    {
+      self->progress = CLAMP (curr / full, 0, 1);
+    }
+
+  g_object_notify_by_pspec (G_OBJECT (self), props[PROP_PROGRESS]);
+  return G_SOURCE_CONTINUE;
+}
+
+static void
+gis_scribe_gpg_watch (GPid     pid,
+                      gint     status,
+                      gpointer data)
+{
+  g_autoptr(GTask) task = G_TASK (data);
+  GisScribe *self = GIS_SCRIBE (g_task_get_source_object (task));
+
+  g_spawn_close_pid (pid);
+
+  g_source_remove (self->gpg_watch);
+  self->gpg_watch = 0;
+
+  g_io_channel_shutdown (self->gpg_stdout, TRUE, NULL);
+  g_clear_pointer (&self->gpg_stdout, g_io_channel_unref);
+
+  if (g_spawn_check_exit_status (status, NULL))
+    {
+      g_task_return_boolean (task, TRUE);
+      // TODO
+    }
+  else
+    {
+      g_task_return_new_error (task, GIS_INSTALL_ERROR, 0,
+                               _("Image verification error."));
+    }
+}
+
+static void
+gis_scribe_begin_verify (GisScribe *self,
+                         GTask     *task)
+{
+  g_autofree gchar *image_path = g_file_get_path (self->image);
+  g_autofree gchar *signature_path = g_file_get_path (self->signature);
+  g_autofree gchar *size_str = g_strdup_printf ("%" G_GUINT64_FORMAT, self->image_size);
+  const gchar * const args[] = {
+      "gpg",
+      "--enable-progress-filter", "--status-fd", "1",
+      /* Trust the one key in this keyring, and no others */
+      "--keyring", IMAGE_KEYRING,
+      "--no-default-keyring",
+      "--trust-model", "always",
+      "--input-size-hint", size_str,
+      "--verify", signature_path, image_path, NULL
+  };
+  gint outfd;
+  GError *error = NULL;
+
+  if (!g_file_query_exists (self->signature, NULL))
+    {
+      g_task_return_new_error (
+          task, GIS_INSTALL_ERROR, 0,
+          _("The Endless OS signature file \"%s\" does not exist."),
+          signature_path);
+      g_object_unref (task);
+      return;
+    }
+
+  if (!g_spawn_async_with_pipes (NULL, (gchar **) args, NULL,
```

Comment:
I do wonder if this wouldn't be better suited by `GSubprocess` and `GOutputStream` for the stdout pipe. It would allow you do avoid using `g_io_channel_unix_new()` and `g_io_add_watch()`, and you'd be able to read the output buffer of the subprocess synchronously inside a `GTask` thread.

Reply:
The next step is to pipe the file into gpg so we can verify and write in parallel. I do indeed intend to use `GSubprocess` for that.

Reply:
I didn't end up pushing this onto a thread but I may do that if I feel keen tomorrow.